Exclude /tmp/bugbug from docker image

PromiseFru opened the pull request:
By installing the `bugbug` library directly from GitHub using `pip`, we eliminate the need to copy the working directory to `/tmp/bugbug` in the container.

Resolves: #970

PromiseFru commented:
@marco-c, we can use a multi-stage build approach to copy only the necessary dependencies and configurations to the final image.

suhaibmujahid commented:
> we can use a multi-stage build approach to copy only the necessary dependencies and configurations to the final image.

It does not seem to be working. Also, some dependencies are missing in the final image.

One option would be to install bugbug as an editable package (i.e., `pip install --editable /path/to/the/repo`). In this way, we will not have two copies. However, I'm not sure if that will cause unexpected behaviour: https://pip.pypa.io/en/stable/topics/local-project-installs/#editable-installs

Another option would be passing the target hash/branch as an argument and installing the package through the git repo URL.

suhaibmujahid reviewed (commented):
> we can use a multi-stage build approach to copy only the necessary dependencies and configurations to the final image.

It does not seem to be working. Also, some dependencies are missing in the final image.

One option would be to install bugbug as an editable package (i.e., `pip install --editable /path/to/the/repo`). In this way, we will not have two copies. However, I'm not sure if that will cause unexpected behaviour: https://pip.pypa.io/en/stable/topics/local-project-installs/#editable-installs

Another option would be passing the target hash/branch as an argument and installing the package through the git repo URL.

marco-c commented:
> Another option would be passing the target hash/branch as an argument and installing the package through the git repo URL.

This might be painful (we'd also need to handle PRs from forks for example)

PromiseFru commented:
> > we can use a multi-stage build approach to copy only the necessary dependencies and configurations to the final image.
> 
> It does not seem to be working. Also, some dependencies are missing in the final image.
> 
> One option would be to install bugbug as an editable package (i.e., `pip install --editable /path/to/the/repo`). In this way, we will not have two copies. However, I'm not sure if that will cause unexpected behaviour: https://pip.pypa.io/en/stable/topics/local-project-installs/#editable-installs
> 
> Another option would be passing the target hash/branch as an argument and installing the package through the git repo URL.

Hello @suhaibmujahid, installing bugbug in `editable` mode didn't remove the duplicate, and the image size remained the same as when installed in `distribution` mode.

```bash
docker build -t mozilla/bugbug-base -f infra/dockerfile.base .
```

```bash
[+] Building 200.0s (11/11) FINISHED                                                                                                                                                       docker:default
 => [internal] load build definition from dockerfile.base                                                                                                                                            0.0s
 => => transferring dockerfile: 677B                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                    0.0s
 => => transferring context: 1.21kB                                                                                                                                                                  0.0s
 => [internal] load metadata for docker.io/library/python:3.10.7-slim                                                                                                                                1.2s
 => [1/6] FROM docker.io/library/python:3.10.7-slim@sha256:f2ee145f3bc4e061f8dfe7e6ebd427a410121495a0bd26e7622136db060c59e0                                                                          7.3s
 => => resolve docker.io/library/python:3.10.7-slim@sha256:f2ee145f3bc4e061f8dfe7e6ebd427a410121495a0bd26e7622136db060c59e0                                                                          0.1s
 => => sha256:bd159e379b3b1bc0134341e4ffdeab5f966ec422ae04818bb69ecef08a823b05 31.42MB / 31.42MB                                                                                                     1.0s
 => => sha256:de08aeb7fd50562d57cef1a49d6197d619df0b4ce52e4caeba2402c27c6e536b 1.08MB / 1.08MB                                                                                                       0.4s
 => => sha256:30527e10f55af538c1464ce755f9e9fceae68bb639f9b8211cfaf3c11e83ba42 12.11MB / 12.11MB                                                                                                     0.6s
 => => sha256:f2ee145f3bc4e061f8dfe7e6ebd427a410121495a0bd26e7622136db060c59e0 1.86kB / 1.86kB                                                                                                       0.0s
 => => sha256:f8fbb2370c6314c806b2ddbec8d94375987e16bc122379bef979c6fc5e962920 1.37kB / 1.37kB                                                                                                       0.0s
 => => sha256:28358ad706063a4fab219d8bcb048124e6dbc6245015b4267b978363af064d42 7.50kB / 7.50kB                                                                                                       0.0s
 => => sha256:693e7a5ba2a8a42880d04042159d67aa9ebb2d67864b7b419fe2847294c080a2 234B / 234B                                                                                                           0.7s
 => => sha256:c7b6f7685fa51adb5da18c9a4fd647bc9cb24af571c32a824206e8a9277d7b09 3.34MB / 3.34MB                                                                                                       0.9s
 => => extracting sha256:bd159e379b3b1bc0134341e4ffdeab5f966ec422ae04818bb69ecef08a823b05                                                                                                            4.0s
 => => extracting sha256:de08aeb7fd50562d57cef1a49d6197d619df0b4ce52e4caeba2402c27c6e536b                                                                                                            0.1s
 => => extracting sha256:30527e10f55af538c1464ce755f9e9fceae68bb639f9b8211cfaf3c11e83ba42                                                                                                            1.0s
 => => extracting sha256:693e7a5ba2a8a42880d04042159d67aa9ebb2d67864b7b419fe2847294c080a2                                                                                                            0.0s
 => => extracting sha256:c7b6f7685fa51adb5da18c9a4fd647bc9cb24af571c32a824206e8a9277d7b09                                                                                                            0.5s
 => [internal] load build context                                                                                                                                                                    2.0s
 => => transferring context: 16.78MB                                                                                                                                                                 2.0s
 => [2/6] COPY requirements.txt /requirements.txt                                                                                                                                                    0.1s
 => [3/6] RUN apt-get update &&  apt-get install -y --no-install-recommends gcc g++ libgomp1 libffi-dev libjemalloc2 zstd patch &&  pip install --disable-pip-version-check --quiet --no-cache-di  168.1s
 => [4/6] COPY infra/mozci_config.toml /root/.config/mozci/config.toml                                                                                                                               0.2s
 => [5/6] COPY . /tmp/bugbug                                                                                                                                                                         1.0s
 => [6/6] RUN pip install --disable-pip-version-check --quiet --no-cache-dir --editable /tmp/bugbug                                                                                                  4.8s
 => exporting to image                                                                                                                                                                              17.1s
 => => exporting layers                                                                                                                                                                             17.1s
 => => writing image sha256:7183ebcb358d28275d5df42d8dd5f37fac63d10b137fea6f86337c7d992c53f0                                                                                                         0.0s
 => => naming to docker.io/mozilla/bugbug-base       
```

```bash
docker images
```

```bash
REPOSITORY               TAG       IMAGE ID       CREATED         SIZE
mozilla/bugbug-base      latest    7183ebcb358d   7 minutes ago   1.5GB
```

```bash
docker run --rm -it mozilla/bugbug-base bash
find . -type d -name bugbug
```

```bash
./usr/local/lib/python3.10/site-packages/mozci/data/sources/bugbug
./tmp/bugbug
./tmp/bugbug/bugbug
./tmp/bugbug/.mypy_cache/3.10/bugbug
```

However, we could use Docker [mount types](https://docs.docker.com/engine/reference/builder/#mount-types) to bind the stages of the multi-build Docker file. The first stage would be just to copy the repository into the container. This way, no dependencies can be omitted in the final stage of the build. We can then bind the repository from the initial stage for installation. With this approach, the image size is reduced, and the image doesn't have a duplicate of the source code.

```bash
docker build -t mozilla/bugbug-base -f infra/dockerfile.base .
```

```bash
[+] Building 213.8s (11/11) FINISHED                                                                                 docker:default
 => [internal] load build definition from dockerfile.base                                                                      0.0s
 => => transferring dockerfile: 768B                                                                                           0.0s
 => [internal] load .dockerignore                                                                                              0.1s
 => => transferring context: 1.21kB                                                                                            0.0s
 => [internal] load metadata for docker.io/library/python:3.10.7-slim                                                          2.0s
 => [internal] load build context                                                                                              1.2s
 => => transferring context: 16.78MB                                                                                           1.2s
 => [builder 1/2] FROM docker.io/library/python:3.10.7-slim@sha256:f2ee145f3bc4e061f8dfe7e6ebd427a410121495a0bd26e7622136db06  5.0s
 => => resolve docker.io/library/python:3.10.7-slim@sha256:f2ee145f3bc4e061f8dfe7e6ebd427a410121495a0bd26e7622136db060c59e0    0.0s
 => => sha256:f2ee145f3bc4e061f8dfe7e6ebd427a410121495a0bd26e7622136db060c59e0 1.86kB / 1.86kB                                 0.0s
 => => sha256:f8fbb2370c6314c806b2ddbec8d94375987e16bc122379bef979c6fc5e962920 1.37kB / 1.37kB                                 0.0s
 => => sha256:28358ad706063a4fab219d8bcb048124e6dbc6245015b4267b978363af064d42 7.50kB / 7.50kB                                 0.0s
 => => sha256:bd159e379b3b1bc0134341e4ffdeab5f966ec422ae04818bb69ecef08a823b05 31.42MB / 31.42MB                               1.0s
 => => sha256:de08aeb7fd50562d57cef1a49d6197d619df0b4ce52e4caeba2402c27c6e536b 1.08MB / 1.08MB                                 0.2s
 => => sha256:30527e10f55af538c1464ce755f9e9fceae68bb639f9b8211cfaf3c11e83ba42 12.11MB / 12.11MB                               0.5s
 => => sha256:693e7a5ba2a8a42880d04042159d67aa9ebb2d67864b7b419fe2847294c080a2 234B / 234B                                     0.5s
 => => sha256:c7b6f7685fa51adb5da18c9a4fd647bc9cb24af571c32a824206e8a9277d7b09 3.34MB / 3.34MB                                 0.8s
 => => extracting sha256:bd159e379b3b1bc0134341e4ffdeab5f966ec422ae04818bb69ecef08a823b05                                      2.5s
 => => extracting sha256:de08aeb7fd50562d57cef1a49d6197d619df0b4ce52e4caeba2402c27c6e536b                                      0.1s
 => => extracting sha256:30527e10f55af538c1464ce755f9e9fceae68bb639f9b8211cfaf3c11e83ba42                                      0.7s
 => => extracting sha256:693e7a5ba2a8a42880d04042159d67aa9ebb2d67864b7b419fe2847294c080a2                                      0.0s
 => => extracting sha256:c7b6f7685fa51adb5da18c9a4fd647bc9cb24af571c32a824206e8a9277d7b09                                      0.4s
 => [stage-1 2/5] COPY requirements.txt /requirements.txt                                                                      0.1s
 => [builder 2/2] COPY . /tmp/bugbug                                                                                           0.5s
 => [stage-1 3/5] RUN apt-get update &&  apt-get install -y --no-install-recommends gcc g++ libgomp1 libffi-dev libjemalloc  183.0s
 => [stage-1 4/5] COPY infra/mozci_config.toml /root/.config/mozci/config.toml                                                 0.2s
 => [stage-1 5/5] RUN --mount=target=/tmp/bugbug,from=builder,source=/tmp/bugbug,rw  pip install --disable-pip-version-check   5.2s
 => exporting to image                                                                                                        18.1s
 => => exporting layers                                                                                                       18.1s
 => => writing image sha256:dcb097723c74ce26ead01c6e4c4c934702543f4de090e0cc59389ebdf43b064d                                   0.0s
 => => naming to docker.io/mozilla/bugbug-base      
```

```bash
docker images
```

```bash
REPOSITORY               TAG       IMAGE ID       CREATED         SIZE
mozilla/bugbug-base      latest    dcb097723c74   53 seconds ago   1.49GB
```

```bash
docker run --rm -it mozilla/bugbug-base bash
find . -type d -name bugbug
```

```bash
./usr/local/lib/python3.10/site-packages/bugbug
./usr/local/lib/python3.10/site-packages/mozci/data/sources/bugbug
```

suhaibmujahid commented on infra/dockerfile.base:
```diff
@@ -1,17 +1,21 @@
 
-COPY . /tmp/bugbug
-RUN pip install --disable-pip-version-check --quiet --no-cache-dir /tmp/bugbug
+RUN --mount=target=/tmp/bugbug,from=builder,source=/tmp/bugbug,rw \
+	pip install --disable-pip-version-check --quiet --no-cache-dir /tmp/bugbug
```
Nice solution! We could use it directly without having a multi-stage build or doing any `COPY`. See: https://docs.docker.com/build/guide/mounts/#add-bind-mounts

PromiseFru commented:
Hello, @suhaibmujahid, thank you for your review. Based on your feedback, I implemented the Docker mount without the multi-stage build or COPY. I also removed the COPY from the `requirement.txt` and used the bind approach since the file isn't required in the image at runtime. I wasn't sure if `infra/mozci_config.toml` is required at runtime, so I left that unchanged.

suhaibmujahid commented on infra/dockerfile.base:
```diff
@@ -1,17 +1,17 @@
-		apt-get purge -y gcc g++ libffi-dev patch && \
-		apt-get autoremove -y && \
-		rm -rf /var/lib/apt/lists/*
+RUN --mount=type=bind,source=requirements.txt,target=/requirements.txt,rw \
```
why do we need `rw`?
```suggestion
RUN --mount=type=bind,source=requirements.txt,target=/requirements.txt\
```

PromiseFru replied:
Ah right! This isn't relevant here. I'll scrap it 🙂

suhaibmujahid commented on infra/dockerfile.base:
```diff
@@ -1,17 +1,17 @@
 
-COPY . /tmp/bugbug
-RUN pip install --disable-pip-version-check --quiet --no-cache-dir /tmp/bugbug
+RUN --mount=type=bind,target=/tmp/bugbug,rw \
```
```suggestion
RUN --mount=type=bind,target=/tmp/bugbug \
```

PromiseFru replied:
Without `rw`, the build process fails with the following error:

```
5.767       error: could not create 'build': Read-only file system
```

According to this [thread](https://discuss.python.org/t/installing-from-a-read-only-source/4115), pip requires write access to build the wheel in the source directory. This issue persists unless you use `--use-deprecated=out-of-tree-build`, which is planned to be removed in pip version 22.1 as mentioned in this [GitHub pull request](https://github.com/pypa/pip/pull/10495/files).

suhaibmujahid replied:
Thank you for the clarification.

suhaibmujahid reviewed (approved):
LGTM! We could file a follow-up issue to merge the two `RUN` instructions into one.
